### ISSUE TITLE
unarchive does not recognize size differs

### ISSUE THREAD

**Issue by ansibot** [ansible/ansible#29610]
From @dknopp on 2016-10-04T14:20:04Z

##### ISSUE TYPE

- Feature Idea
##### COMPONENT NAME

unarchive
##### ANSIBLE VERSION

```
ansible 2.1.1.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
```
##### CONFIGURATION

Nothing unusual or special about configuration, simply using defaults.
##### OS / ENVIRONMENT

CentOS 6.6 32bit
##### SUMMARY

The unarchive task does not consider `Size differs` in the output of gtar in it's is_unarchived method, which perhaps was by initial design or just not thought of. 

As an enhancement it would be nice to add control over if `Size differs` is considered. 

The situation that this would be used for is when you have a custom built archive that needs to replace existing files, for example: configuration or python scripts or static website files/resources.

I did see a ticket on adding a checksum option so perhaps the above situation could be resolved using that feature instead, in which case this is a duplicate or won't fix.

Finally perhaps using the force parameter as a way to always unarchive would also be a solution. This doesn't seem to be the current case, if it is then it should be documented.

Here is a snippet from the unarchive.py is_unarchived method and the output of gtar --diff when the file size differs.

```
[root@niu01 v-ui-crs]# gtar --diff -vf /opt/xpressworkx/app-manager/apps/web/application-files.tar application

application/
application/__init__.py: Size differs
application/configure.py: Mod time differs
application/configure.py: Size differs
application/override.py: Size differs
```

From /usr/lib/python2.6/site-packages/ansible/modules/core/files/unarchive.py 

```
    def is_unarchived(self):
        cmd = '%s -C "%s" -d%s' % (self.cmd_path, self.dest, self.zipflag)
        if self.opts:
            cmd += ' ' + ' '.join(self.opts)
        if self.file_args['owner']:
            cmd += ' --owner="%s"' % self.file_args['owner']
        if self.file_args['group']:
            cmd += ' --group="%s"' % self.file_args['group']
        if self.file_args['mode']:
            cmd += ' --mode="%s"' % self.file_args['mode']
        if self.module.params['keep_newer']:
            cmd += ' --keep-newer-files'
        if self.excludes:
            cmd += ' --exclude="' + '" --exclude="'.join(self.excludes) + '"'
        cmd += ' -f "%s"' % self.src
        rc, out, err = self.module.run_command(cmd)

        ....
        ....

```
##### STEPS TO REPRODUCE

Simply use an unarchive task in a situation where you need to replace existing files of the same name, owner, group, and mode.

```
- name: Unpack Static Files
  unarchive:
    src: "{{ role_path }}/../../web/sdist/static-files.tar"
    dest: "{{ web_config.data['file_root'] }}"
    owner: "{{ web_config.data['web_user'] }}"

```

##### EXPECTED RESULTS

Unarchive the tar into the destination even though files of the same name, owner, group and mode exist.
##### ACTUAL RESULTS

Task returned ok as in already run.

```
TASK [web-ui-common : Unpack Application Static Files] *************************
ok: [192.168.124.152]
```

Copied from original issue: ansible/ansible-modules-core#5149

**Comment by ansibot** [comment]
From @ansibot on 2016-10-04T14:20:04Z



@dagwieers, @pileofrogs, ping. This issue is waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
From @dagwieers on 2016-10-04T14:20:04Z

This is indeed a bug, we don't consider size changes at this point. I just added a fix to an existing open PR #5721 

Can you check if this change fixes your issue ? (see the SIZE_DIFF_RE pieces)

**Comment by mishas** [comment]
Is there any update on this?
It seems like there was a fix in the previous repository (before migration): https://github.com/ansible/ansible-modules-core/pull/5721/files#diff-c8620ce6bed12fc5110958c7607f1b03R161,
but it's not here after the migration (and also the PR referenced above (#5721) is incorrectly linked (it was the PR ID in the old repository).
